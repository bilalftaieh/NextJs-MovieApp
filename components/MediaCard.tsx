'use client'
// Importing necessary libraries
import React from 'react';
import NextImage from 'next/image';
import Link from 'next/link';
import { formatDateString } from '@/lib/utils';
import { Card, CardBody, CardFooter, Image } from "@nextui-org/react";

// Props for MediaCard component
type MediaCardProps = {
    children?: React.ReactNode; // Children of the MediaCard component
    imageUrl: string; // URL of the image
    mediaName?: string; // Name of the media
    releaseDate?: string; // Release date of the movie
    className?: String; // Class name for the MediaCard component
    href?: string; // Href for the MediaCard component
    tagline?: string; // Tagline for the MediaCard component
};

// The main function for the MediaCard component
const MediaCard: React.FC<MediaCardProps> = ({ imageUrl, mediaName, releaseDate, href, tagline }) => {
    const hoverScaleClass = href ? 'active:scale-110 md:hover:scale-110' : '';
    // const cardPressableProps = href ? { isPressable: true, onPress: () => push(href) } : {};
    return (
        <Link href={href || ''}>
            <Card className={`bg-transparent max-w-[200px] flex flex-col ${hoverScaleClass} 
            shadow-none`}>
                <CardBody className='flex-grow'>
                    <Image
                        unoptimized
                        as={NextImage}
                        width={200}
                        height={200}
                        alt={mediaName}
                        className="object-cover rounded-xl"
                        priority
                        src={imageUrl}
                        fallbackSrc={'/default-image.png'}
                    />
                </CardBody>
                <CardFooter className="flex flex-col gap-4 ">
                    {mediaName && <p className='text-lg font-bold line-clamp-1'>{mediaName}</p>}
                    {releaseDate && (
                        <p className="text-sm text-custom-two">
                            Release Date: <span className="font-semibold whitespace-nowrap">{formatDateString(releaseDate)}</span>
                        </p>
                    )} {/* Release date of the movie */}
                    {tagline && <p className="text-sm font-semibold italic 
                text-custom-two text-center">{tagline.toUpperCase()}</p>} {/* Tagline for the movie*/}
                </CardFooter>
            </Card>
        </Link>

    )

};

export default MediaCard;
