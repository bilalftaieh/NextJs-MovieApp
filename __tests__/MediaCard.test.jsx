import '@testing-library/jest-dom'
import { render,screen,fireEvent} from '@testing-library/react';
import MediaCard from '@/components/MediaCard';

const mockProps = {
  brokenImageUrl: 'https://path/to/image',
  workingImage : '/dark-knight.jpg',
  href : '/testHref',
  mediaName: 'Test Media',
  releaseDate: '2023-01-01',
  tagline:'Test Tagline'
};

describe('MediaCard', () => {
  it('renders without crashing', () => {
    render(<MediaCard imageUrl={mockProps.workingImage} />);
  });

  it('renders with correct image source', () => {
    render(<MediaCard imageUrl={mockProps.workingImage} />);
    expect(screen.getByAltText('No Image').getAttribute('src')).toContain('dark-knight.jpg');
  });

  it('replaces image source on error', () => {
    render(<MediaCard imageUrl={mockProps.brokenImageUrl} />);
    const img = screen.getByRole('img');

    // Simulate an error on the image
    fireEvent.error(img);

    // Check if the image source has been replaced
    expect(img.src).toMatch('default-image.png');
});

  it('renders with correct movie name', () => {
    render(<MediaCard imageUrl={mockProps.workingImage} mediaName={mockProps.mediaName} />);
    expect(screen.getByText(`${mockProps.mediaName}`)).toBeInTheDocument();
  });

  it('renders with correct release date', () => {
    render(<MediaCard imageUrl={mockProps.workingImage} releaseDate={mockProps.releaseDate} />);
    expect(screen.getByText(`01/01/2023`)).toBeInTheDocument();
  });

  it('renders with correct tagline', () => {
    render(<MediaCard imageUrl={mockProps.workingImage} tagline={mockProps.tagline} />);
    expect(screen.getByText(`${mockProps.tagline}`)).toBeInTheDocument();
  });

  it('renders content wrapped with next/link when href is provided', () => {
    const { container} = render(
        <MediaCard imageUrl={mockProps.workingImage} 
        mediaName={mockProps.mediaName} href={mockProps.href} />
    );

    // Check if the link is rendered
    const linkElement = container.querySelector('a');
    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute('href',`${mockProps.href}`);

    // Check if the content is rendered within the link
    const contentElement = screen.getByText(`${mockProps.mediaName}`);
    expect(linkElement).toContainElement(contentElement);
});
it('matches snapshot', () => {
    const { asFragment, rerender } = render(<MediaCard imageUrl={mockProps.workingImage} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<MediaCard imageUrl={mockProps.workingImage} href={mockProps.href} />);
    expect(asFragment()).toMatchSnapshot();
});

});
