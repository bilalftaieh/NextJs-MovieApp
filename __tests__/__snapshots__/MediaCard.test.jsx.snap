// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="media-card flex flex-col items-center max-w-xs rounded overflow-hidden hover:scale-110 transition-transform cursor-pointer"
  >
    <div
      class="relative h-64 w-full"
    >
      <img
        alt="No Image"
        class="w-auto"
        data-nimg="fill"
        decoding="async"
        fetchpriority="high"
        sizes="100%"
        src="http://localhost/_next/image?url=%2Fdark-knight.jpg&w=3840&q=75"
        srcset="/_next/image?url=%2Fdark-knight.jpg&w=16&q=75 16w, /_next/image?url=%2Fdark-knight.jpg&w=32&q=75 32w, /_next/image?url=%2Fdark-knight.jpg&w=48&q=75 48w, /_next/image?url=%2Fdark-knight.jpg&w=64&q=75 64w, /_next/image?url=%2Fdark-knight.jpg&w=96&q=75 96w, /_next/image?url=%2Fdark-knight.jpg&w=128&q=75 128w, /_next/image?url=%2Fdark-knight.jpg&w=256&q=75 256w, /_next/image?url=%2Fdark-knight.jpg&w=384&q=75 384w, /_next/image?url=%2Fdark-knight.jpg&w=640&q=75 640w, /_next/image?url=%2Fdark-knight.jpg&w=750&q=75 750w, /_next/image?url=%2Fdark-knight.jpg&w=828&q=75 828w, /_next/image?url=%2Fdark-knight.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fdark-knight.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fdark-knight.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fdark-knight.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fdark-knight.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent; background-size: contain; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="flex-grow px-6 py-4 text-center "
    >
         
    </div>
  </div>
</DocumentFragment>
`;

exports[`MediaCard matches snapshot 2`] = `
<DocumentFragment>
  <a
    href="/testHref"
  >
    <div
      class="media-card flex flex-col items-center max-w-xs rounded overflow-hidden hover:scale-110 transition-transform cursor-pointer"
    >
      <div
        class="relative h-64 w-full"
      >
        <img
          alt="No Image"
          class="w-auto"
          data-nimg="fill"
          decoding="async"
          fetchpriority="high"
          sizes="100%"
          src="http://localhost/_next/image?url=%2Fdark-knight.jpg&w=3840&q=75"
          srcset="/_next/image?url=%2Fdark-knight.jpg&w=16&q=75 16w, /_next/image?url=%2Fdark-knight.jpg&w=32&q=75 32w, /_next/image?url=%2Fdark-knight.jpg&w=48&q=75 48w, /_next/image?url=%2Fdark-knight.jpg&w=64&q=75 64w, /_next/image?url=%2Fdark-knight.jpg&w=96&q=75 96w, /_next/image?url=%2Fdark-knight.jpg&w=128&q=75 128w, /_next/image?url=%2Fdark-knight.jpg&w=256&q=75 256w, /_next/image?url=%2Fdark-knight.jpg&w=384&q=75 384w, /_next/image?url=%2Fdark-knight.jpg&w=640&q=75 640w, /_next/image?url=%2Fdark-knight.jpg&w=750&q=75 750w, /_next/image?url=%2Fdark-knight.jpg&w=828&q=75 828w, /_next/image?url=%2Fdark-knight.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fdark-knight.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fdark-knight.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fdark-knight.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fdark-knight.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent; background-size: contain; background-position: 50% 50%; background-repeat: no-repeat;"
        />
      </div>
      <div
        class="flex-grow px-6 py-4 text-center "
      >
           
      </div>
    </div>
  </a>
</DocumentFragment>
`;
